### PR TITLE
Add new test service api to return label values for filter: Fixes #1767

### LINKED ISSUE #1767: UI dataset filter returing All label values for filtering, instead of unique Filtering label values
**Describe the bug**

The call to Test Label Values used in filtering datasets is returning *all* labelValues for a test, instead of a list of **unique filtering**  label values.

image

### PR DESCRIPTION
**Pull request by johnaohara**
## Fixes Issue

Fixes #1767

image

### CODE REVIEW

**Review by lampajr — CHANGES_REQUESTED** [review]
Overall looks great, I added some changes that are more for simplification

File: horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java [hunk 1]
```diff
@@ -103,6 +104,16 @@ SELECT DISTINCT COALESCE(jsonb_object_agg(label.name, lv.value) FILTER (WHERE la
          ) select * from combined FILTER_PLACEHOLDER ORDER_PLACEHOLDER limit :limit offset :offset
          """;
 
+   protected static final String LABEL_VALUES_SUMMARY_QUERY = """
+         SELECT DISTINCT COALESCE(jsonb_object_agg(label.name, lv.value), '{}'::::jsonb) AS values
+                  FROM dataset
+                  INNER JOIN label_values lv ON dataset.id = lv.dataset_id
+                  INNER JOIN label ON label.id = lv.label_id
+                  WHERE dataset.testid = :testId
+                     AND (label.id IS NULL OR (TRUE AND label.filtering))
```
[the remark this answers is not in the document]
Comment (johnaohara) [reply horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:113]: This part of the WHERE clause is not correct now, I will fix it

File: horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java [hunk 2]
```diff
@@ -821,6 +832,33 @@ public List<ExportedLabelValues> labelValues(int testId, String filter, String b
...
+      String sql = LABEL_VALUES_SUMMARY_QUERY;
+
+      NativeQuery query =  ((NativeQuery) em.createNativeQuery(sql))
+              .setParameter("testId", test.id);
+
+      query
+              .unwrap(NativeQuery.class)
+              .addScalar("values", JsonBinaryType.INSTANCE);
+      List<ObjectNode> filters = query.getResultList();
+      if(filters == null || filters.isEmpty())
```
[the remark this answers is not in the document]
Comment (johnaohara) [reply horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:855]: good catch, thanks

File: horreum-web/src/domain/runs/TestDatasets.tsx [hunk 3]
```diff
@@ -205,19 +205,10 @@ export default function TestDatasets() {
...
-        return resultArr;
-    }
 
     const labelsSource = useCallback(() => {
-        return testApi.labelValues(testIdInt)
-            .then((result: Array<ExportedLabelValues>) => {
-                return flattenLabelValues(result);
-            })
+        return testApi.filteringLabelValues(testIdInt)
+            .then((result: Array<any>) => result)
```
[the remark this answers is not in the document]
Comment (johnaohara) [reply horreum-web/src/domain/runs/TestDatasets.tsx:211]: Thanks, have re-factored this further

**Comment by johnaohara** [comment]
@lampajr please can you take another look?

File: horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java [hunk 4]
```diff
@@ -821,6 +831,29 @@ public List<ExportedLabelValues> labelValues(int testId, String filter, String b
...
+   public List<ObjectNode> filteringLabelValues(
+           @PathParam("id") int testId){
+
+      Test test = get(testId,null);
+      if(test == null){
+         throw ServiceException.serverError("Cannot find test "+testId);
+      }
+
+      NativeQuery query =  ((NativeQuery) em.createNativeQuery(LABEL_VALUES_SUMMARY_QUERY))
+              .setParameter("testId", test.id);
```
[the remark this answers is not in the document]
Comment (lampajr) [reply horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:846]: This is actually just a simplification, it does not change the behavior so up to you :rocket:

File: horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java [hunk 5]
```diff
@@ -103,6 +104,15 @@ SELECT DISTINCT COALESCE(jsonb_object_agg(label.name, lv.value) FILTER (WHERE la
          ) select * from combined FILTER_PLACEHOLDER ORDER_PLACEHOLDER limit :limit offset :offset
          """;
 
+   protected static final String LABEL_VALUES_SUMMARY_QUERY = """
+         SELECT DISTINCT COALESCE(jsonb_object_agg(label.name, lv.value), '{}'::::jsonb) AS values
```
Comment (lampajr) [inline horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/TestServiceImpl.java:108]: As you highlighted here https://github.com/Hyperfoil/Horreum/issues/1766 I think you can use `::` and it should work as well
```suggestion
         SELECT DISTINCT COALESCE(jsonb_object_agg(label.name, lv.value), '{}'::jsonb) AS values
```

**Comment by johnaohara** [comment]
@lampajr and again... ;)

**Review by lampajr — APPROVED** [review]
LGTM!

Thanks @johnaohara, sorry for the nit-picking :smiley:

**Pull request merged**